src/main/scala/com/databricks/spark/csv/CsvRelation.scala

```diff
@@ -108,7 +108,14 @@ case class CsvRelation protected[spark] (
         try {
           index = 0
           while (index < schemaFields.length) {
-            rowArray(index) = TypeCast.castTo(tokens(index), schemaFields(index).dataType)
+            rowArray(index) = if (schemaFields(index).nullable && tokens(index) == ""){
```

Comment:
I don't think we need to check for nullability for StringType. Basically I think the following logic would be simple an intuitive:
* if StringType just return the token (don't need to check for nullability)
* if any other time, if token is "", return null, else cast.

I suggest moving this logic (along with some comment that explains it) to TypeCast.castTo (maybe as a simple private method). This way you can add a few simple unit tests for it.